### PR TITLE
toolkit: add text to speech synthesis

### PR DESCRIPTION
**Pull request by danylo-boiko**
**Description**
- Integrated a text-to-speech synthesizer to enable voicing of chatbot responses

**AI Description**

This PR introduces a new feature for synthesizing audio from text messages in conversations. It adds a new endpoint to the CohereClient class, allowing users to synthesize messages by providing conversation and message IDs. The PR also includes changes to the MessageRow component, adding a new property `isSynthesisPlaying` and a new function `onToggleSynthesis` to handle the synthesis of messages.

The PR makes the following changes:

- Adds a new `synthesize_message` function to the `conversation.py` file, which generates a synthesized audio stream for a specific message in a conversation.
- Modifies the `StreamEnd` class in `chat.py` to include a `message_id` field.
- Updates the `generate_chat_stream` function in `chat.py` to include the `message_id` of the response message in the `stream_end_data` dictionary.
- Adds a new `synthesize_stream` function to the `synthesizer.py` file, which generates a stream of audio bytes from the given text using text-to-speech synthesis.
- Adds a new `detect_language` function to the `synthesizer.py` file, which detects the language of the given text.
- Adds a new `useSynthesizer` hook in `use-synthesizer.ts` to manage the synthesis of messages.
- Adds a new `useSynthesizer` hook in `use-synthesizer.ts` to manage the synthesis of messages.
- Adds new icons for 'Stop' and 'Volume' in `Stop.tsx` and `Volume.tsx` respectively.
- Exports the new icons from `index.ts`.
- Adds a new `synthesizeMessage` function to the `CohereClient` class in `client.ts`, which fetches the synthesized message from the server.
- Adds a new `message_id` field to the `$StreamEnd` object in `schemas.gen.ts`.
- Adds a new `message_id` field to the `StreamEnd` type in `types.gen.ts`.
- Adds a new `id` field to the `BaseMessage` type in `message.ts`.
- Adds a new `id` field to the `message` object in `conversation.ts`.

### CODE REVIEW

**Comment by danylo-boiko** [comment]
Hey @tianjing-li! Could you please take a look?

**Comment by tianjing-li** [comment]
Hey @danylo-boiko thanks for this contribution - @malexw and I shared some concerns over the libraries used - 

They both seem to use undocumented Google APIs which could break at any moment, are there any alternatives to this? If not, you can just confirm with me and I'll have to confirm internally that this is a direction we want to take

**Comment by tianjing-li** [comment]
@danylo-boiko I discussed this internally, would it be possible for you to replace the two libraries with Google Cloud APIs? It would require an API key and the APIs enabled in a Google Cloud project but ultimately this is a direction we'd be more comfortable taking.

**Comment by danylo-boiko** [comment]
> @danylo-boiko I discussed this internally, would it be possible for you to replace the two libraries with Google Cloud APIs? It would require an API key and the APIs enabled in a Google Cloud project but ultimately this is a direction we'd be more comfortable taking.

Of course, I'll let you know

**Comment by tianjing-li** [comment]
@danylo-boiko Thank you for the flexibility, happy to re-review this feature with those adjustments

**Comment by danylo-boiko** [comment]
Hello hello @tianjing-li. Google Drive related changes are ready, please check it out.

**Comment by tianjing-li** [comment]
@danylo-boiko  thank you! Will check it out today

File: src/backend/services/synthesizer.py
Comment (tianjing-li) [inline src/backend/services/synthesizer.py:8]: Thanks for switching to the Cloud APIs

File: src/backend/routers/conversation.py
Comment (tianjing-li) [inline src/backend/routers/conversation.py:583]: Let's add some sort of validation in the scenario where `synthesize` fails. Wrapping it in a try/catch would be fine

File: src/backend/services/synthesizer.py
Comment (tianjing-li) [inline src/backend/services/synthesizer.py:23]: For both `synthesize` and `detect_language`, let's add some validation checks to verify if the API key is properly set. You can throw an error here that the try/catch in the endpoint catches

File: src/backend/services/synthesizer.py
Comment (tianjing-li) [inline src/backend/services/synthesizer.py:27]: Nit: `language`

File: src/backend/services/synthesizer.py
Comment (tianjing-li) [inline src/backend/services/synthesizer.py:49]: I don't think this assumption can be made, for example the phrase: `J'aime manger des pommes` is ascii, but is french. Let's remove this if

File: src/backend/services/synthesizer.py
Comment (tianjing-li) [inline src/backend/services/synthesizer.py:51]: Same thing here, I would validate the api key exists

**Review by tianjing-li — CHANGES_REQUESTED** [review]
Some changes requested

File: src/backend/services/synthesizer.py
Comment (danylo-boiko) [reply src/backend/services/synthesizer.py]: Yeah, that's a good point, thanks!

File: src/backend/services/synthesizer.py
Comment (danylo-boiko) [reply src/backend/services/synthesizer.py]: Done

File: src/backend/routers/conversation.py
Comment (danylo-boiko) [reply src/backend/routers/conversation.py]: Done

File: src/backend/services/synthesizer.py
Comment (danylo-boiko) [reply src/backend/services/synthesizer.py:26]: Makes sense, done

**Comment by danylo-boiko** [comment]
> Some changes requested

@tianjing-li thanks for the code review. Please take a look one more time.

**Comment by danylo-boiko** [comment]
@tianjing-li, I also added the setup guide as a separate PR: https://github.com/cohere-ai/cohere-toolkit/pull/812

**Review by tianjing-li — APPROVED** [review]
Lgtm!

**Comment by tianjing-li** [comment]
@danylo-boiko approved! Once the checks complete I'll merge it today

**Comment by tianjing-li** [comment]
@danylo-boiko There seems to be a minor typechecking error, you can run pyright on your `synthesizer.py` file locally to see, the error is:

`/home/runner/work/cohere-toolkit/cohere-toolkit/src/backend/services/synthesizer.py:71:39 - error: "api_key" is not a known member of "None" (reportOptionalMemberAccess)`

**Comment by danylo-boiko** [comment]
> @danylo-boiko There seems to be a minor typechecking error, you can run pyright on your `synthesizer.py` file locally to see, the error is:
> 
> `/home/runner/work/cohere-toolkit/cohere-toolkit/src/backend/services/synthesizer.py:71:39 - error: "api_key" is not a known member of "None" (reportOptionalMemberAccess)`

Fixed, thanks!

**Comment by tianjing-li** [comment]
@danylo-boiko all checks passing now - merging

**Pull request merged**
